### PR TITLE
Serve static assets instead of using CDN

### PR DESCRIPTION
**Pull request by sbs2001**
Fixes https://github.com/google/vulncode-db/issues/31

The versions of static assets are unchanged.

Included all `*.map` files too but not sure whether we want them here. Please let me know. 

Signed-off-by: Shivam Sandbhor <shivam.sandbhor@gmail.com>

### CODE REVIEW

**Comment by sbs2001** [comment]
@googlebot I signed it!

**Comment by sbs2001** [comment]
@evonide  Why are is the checkout step failing in the tests ? Tests are successful on my local machine via `./docker/docker-admin.sh test`

**Comment by evonide** [comment]
Thanks a lot for the commit Shivam!
My colleague or I will look into it as soon as possible and get back with feedback here soon.

**Comment by bluec0re** [comment]
Thanks for the changes :). Looks like github workflow is (was) flawed.

Could you quickly point me to where you got the following files from?

- monaco/min/vs/loader.js
- all.min.css
- showdown.min.js
- bootstrap.min.js
- bootstrap.min.css
- mdb.css

I couldn't verify that they are the same as before:
```
$ sha256sum -c ../cdn_files.sha256 
static/monaco/min/vs/loader.js: FAILED
static/css/third_party/style.min.css: OK
static/css/third_party/jquery.dataTables.min.css: OK
static/css/third_party/all.min.css: FAILED
static/css/third_party/jquery-ui.min.css: OK
static/js/third_party/jquery.min.js: OK
static/js/third_party/jquery.dataTables.min.js: OK
static/js/third_party/jstree.min.js: OK
static/js/third_party/jquery-ui.min.js: OK
static/js/third_party/showdown.min.js: FAILED
sha256sum: WARNING: 3 computed checksums did NOT match
$ sha384sum -c ../cdn_files.sha384 
static/js/third_party/popper.min.js: OK
static/js/third_party/bootstrap.min.js: FAILED
static/css/third_party/bootstrap.min.css: FAILED
sha384sum: WARNING: 2 computed checksums did NOT match
$ sha512sum -c ../cdn_files.sha512
static/js/third_party/mdb.min.js: OK
static/css/third_party/mdb.css: FAILED
sha512sum: WARNING: 1 computed checksum did NOT match
```

Thanks again!

**Comment by sbs2001** [comment]
@bluec0re  Thanks for the review. 

I got all the `monaco` assets by installing https://www.npmjs.com/package/monaco-editor/v/0.20.0 .  After a second a look it looks like the hashes failed because I've used `monaco v0.20.0` rather than the currently used `monaco v0.14.3`.

Font awesome's `all.min.css` is  obtained from  https://github.com/FortAwesome/Font-Awesome/blob/5.14.0/css/all.min.css

As for `showdown.min.js` I  think the hash failed due because I had removed the `map` . I have added a commit and hopefully the hash will match. I've used `https://raw.githubusercontent.com/showdownjs/showdown/1.8.7/dist/showdown.min.js`.

All bootstrap files are from `https://github.com/twbs/bootstrap/releases/download/v4.2.1/bootstrap-4.2.1-dist.zip`

`mdb.css` has the same issue like `showdown.min.js`, I've added a commit to fix it. The hashes will tally up hopefully.

**Comment by sbs2001** [comment]
@bluec0re tests are passing thanks for  https://github.com/google/vulncode-db/commit/5eaa255b3eae04038b999cff03e12cf5b6c5ea49

File: templates/footer.html [hunk 1]
```diff
@@ -56,9 +48,9 @@ <h5 class="mb-1">Footer</h5>
...
       return `data:text/javascript;charset=utf-8,${encodeURIComponent(`
         self.MonacoEnvironment = {
-          baseUrl: 'https://cdnjs.cloudflare.com/ajax/libs/monaco-editor/0.14.3/min/'
+          baseUrl: "{{ url_for('frontend.serve_static', path='monaco/min') }}"
         };
-        importScripts('https://cdnjs.cloudflare.com/ajax/libs/monaco-editor/0.14.3/min/vs/base/worker/workerMain.js');`
+        importScripts("{{ url_for('frontend.serve_static', path='monaco/min/vs/base/worker/workerMain.js') }}");`
       )}`;
     }
   };
```
Comment (bluec0re) [inline templates/footer.html:56]: can you remove the whole `window.MonacoEnviroment` part (including comment). As monaco is now served directly, this is a) not needed anymore and b) breaks some functionality

File: templates/vulnerability/embedded.html [hunk 2]
```diff
@@ -82,9 +75,9 @@
...
       return `data:text/javascript;charset=utf-8,${encodeURIComponent(`
         self.MonacoEnvironment = {
-          baseUrl: 'https://cdnjs.cloudflare.com/ajax/libs/monaco-editor/0.14.3/min/'
+          baseUrl: "{{ url_for('frontend.serve_static', path='monaco/min') }}"
         };
-        importScripts('https://cdnjs.cloudflare.com/ajax/libs/monaco-editor/0.14.3/min/vs/base/worker/workerMain.js');`
+        importScripts("{{ url_for('frontend.serve_static', path='monaco/min/vs/base/worker/workerMain.js') }}");`
       )}`;
     }
   };
```
Comment (bluec0re) [inline templates/vulnerability/embedded.html:83]: can you remove the whole `window.MonacoEnviroment` part (including comment). As monaco is now served directly, this is a) not needed anymore and b) breaks some functionality

**Review by bluec0re — CHANGES_REQUESTED** [review]
The monaco configuration needs some adjustments if it's served from the same origin. I marked the code parts.

Monaco 0.20.0 has also some breaking changes for us compared to the (admittedly) old version. I'll fix them with a follow up change

**Comment by sbs2001** [comment]
@bluec0re  Could you point me to some endpoint which relies on monaco ? 

I'm trying to see what happens if  `monaco`  is not properly set up (so I would know whether my changes are breaking the code or not) but I'm  not successful. I tried removing `<script>` tags related to monaco from `footer.html` and `embedded.html` but I noticed no changes in the UI. I can still propose changes with the same UI whether I include `monaco` or not.

File: templates/footer.html (hunk 1 above)
Comment (sbs2001) [reply templates/footer.html]: I've pushed a commit  implementing the suggestion. I'm not sure whether monaco is properly working.  Please provide me some litmus test to determine whether monaco is working properly (see my comment below).

**Comment by bluec0re** [comment]
Sorry for the delay.

The editor is currently used in multiple places, examples are:

- https://www.vulncode-db.com/CVE-2014-1912
- https://www.vulncode-db.com/CVE-2014-1912/editor
- https://www.vulncode-db.com/CVE-2014-1912/embed

But you have to setup the vcs_proxy and complete the code location part of an entry in your local version to enable it (as you probably already tried).

Don't know why you get a 404 for your review. You followed the link to the entry in the "Reviews" page right?
@evonide who implemented the review mechanism

**Review by bluec0re — APPROVED** [review]
FYI: web workers are still used by monaco, but they are now served from the same domain they are instantiated at, so the same-domain proxy is not required anymore.

**Comment by evonide** [comment]
Thanks so much for this commit Shivam!

Please note that the review process is currently work in progress (https://github.com/google/vulncode-db/issues/14) and might have issues. Particularly, the "propose changes" is only affecting meta data and is not connected to monaco.
To test the editor you might need to run:
```
# Search for entries with patch links and add additional application entries for them.
./docker/docker-admin.sh crawl_patches
```
If you haven't already done so. To annotate code you would then need to:

1. Visit an entry (http://127.0.0.1:8080/[ID]) that has a patch linked as described above.
2. Particularly, click "Annotate entry" in the entry or visit http://127.0.0.1:8080/[ID]/editor (please note as described in https://github.com/google/vulncode-db/pull/42#issuecomment-698158776 you would need to run the vcs_proxy, too).
3. Use the editor there. For example you can right click and create a comment or mark regions and click "Save changes".

Once a comment or marked regions are available http://127.0.0.1:8080/[ID] should directly display the data as can be seen in https://www.vulncode-db.com/CVE-2014-0160.

Both features "annotate entry" and "propose changes" will be merged into one single thing to avoid confusion.
Please let us know if you encounter any difficulties or need other support!

**Pull request merged**